### PR TITLE
Fix buf: if arguments is a sections in SplitAxis

### PR DESCRIPTION
**Pull request by okuta**
I fixed `foward_gpu` function.

### CODE REVIEW

**Comment by delta2323** [comment]
I'll check it

File: chainer/functions/split_axis.py [hunk 1]
```diff
@@ -33,21 +32,20 @@ def forward_gpu(self, x):
...
-        indices = self.indices_or_sections
-        if isinstance(indices, collections.Iterable):
-            indices = list(indices)
-            indices.append(xshape[self.axis])
+        if isinstance(self.indices_or_sections, collections.Iterable):
+            ind = list(self.indices_or_sections)
+            ind.append(xshape[self.axis])
         else:
-            if xshape[self.axis] % indices:
+            sec = self.indices_or_sections
```
Comment (delta2323) [inline chainer/functions/split_axis.py]: It is better to check if `self.dices_or_sections` is integer and raise an error if this assumption is not satisfied.

File: chainer/functions/split_axis.py [hunk 2]
```diff
@@ -33,21 +32,20 @@ def forward_gpu(self, x):
         self.cdimx = xshape[self.axis]
         self.rdim = numpy.prod(xshape[self.axis + 1:])
 
-        indices = self.indices_or_sections
-        if isinstance(indices, collections.Iterable):
-            indices = list(indices)
-            indices.append(xshape[self.axis])
+        if isinstance(self.indices_or_sections, collections.Iterable):
+            ind = list(self.indices_or_sections)
+            ind.append(xshape[self.axis])
```
Comment (delta2323) [inline chainer/functions/split_axis.py]: xshape[self.axis] is assigned to `self.cdimx`.

File: chainer/functions/split_axis.py [hunk 3]
```diff
@@ -33,21 +32,20 @@ def forward_gpu(self, x):
...
         else:
-            if xshape[self.axis] % indices:
+            sec = self.indices_or_sections
+            if xshape[self.axis] % sec:
```
Comment (delta2323) [inline chainer/functions/split_axis.py]: `xshape[self.axis]` is assigned to `self.cdimx`.

File: chainer/functions/split_axis.py [hunk 4]
```diff
@@ -33,21 +32,20 @@ def forward_gpu(self, x):
...
                 raise ValueError(
                     'array split does not result in an equal division')
-            indices = six.moves.range(
-                indices, xshape[self.axis] + indices, indices)
+            ind = numpy.arange(1, sec + 1) * (xshape[self.axis] // sec)
```
Comment (delta2323) [inline chainer/functions/split_axis.py]: `xshape[self.axis]` is assigned to `self.cdimx`.

File: chainer/functions/split_axis.py [hunk 5]
```diff
@@ -33,21 +32,20 @@ def forward_gpu(self, x):
...
         ys = []
         kernel = cuda.elementwise(
             _args, 'COPY(y[i] = x[idx])', 'split_fwd', preamble=_preamble)
         bi = 0
-        for i in indices:
+        for i in ind:
             i = min(i, xshape[self.axis])
```
Comment (delta2323) [inline chainer/functions/split_axis.py]: `xshape[self.axis]` is assigned to `self.cdimx`.

File: chainer/functions/split_axis.py [hunk 6]
```diff
@@ -33,21 +32,20 @@ def forward_gpu(self, x):
...
+            if xshape[self.axis] % sec:
                 raise ValueError(
                     'array split does not result in an equal division')
-            indices = six.moves.range(
-                indices, xshape[self.axis] + indices, indices)
+            ind = numpy.arange(1, sec + 1) * (xshape[self.axis] // sec)
         ys = []
         kernel = cuda.elementwise(
             _args, 'COPY(y[i] = x[idx])', 'split_fwd', preamble=_preamble)
         bi = 0
```
Comment (delta2323) [inline chainer/functions/split_axis.py]: It is better that the meaning of this variable should be clear from its name (like `i_prev`, `prev_i`)

**Comment by delta2323** [comment]
Though it is not related to this PR directly, I found that we should add test case where `indices_or_section` is a list that contains one or more 0's. In that case, the output of forward is # of non-zero elements in `indices_or_sections` (i.e. tensors that corresponding to 0 are removed). I do not say it is not intuitive. But I think it is not obvious from function's name.

**Comment by okuta** [comment]
The numpy supports zero size array.
But, pycuda partial supports zero size array.
`chainer.cuda.empty((1,0,1))` -> `array([], shape=(1, 0, 1), dtype=float32)`
`chainer.cuda.empty((1,0,1)) + 1` -> error

Therefore, I did not add test cases where indices_or_sections is a list that contains one or more 0's.

File: chainer/functions/split_axis.py [hunk 7]
```diff
@@ -22,10 +21,23 @@ class SplitAxis(function.Function):
...
     def forward_cpu(self, x):
+        if isinstance(self.indices_or_sections, collections.Iterable):
+            cdimx = x[0].shape[self.axis]
+            ind = list(self.indices_or_sections)
+            ind.append(cdimx)
+            prev_i = 0
+            for i in ind:
+                cdimy = max(0, min(i, cdimx) - prev_i)
+                if cdimy == 0:
+                    raise ValueError('Not support if shape contains 0')
```
Comment (delta2323) [inline chainer/functions/split_axis.py]: More explanatory error is appropriate.
Output tensors whose shapes contain 0 are not supported.

File: chainer/functions/split_axis.py [hunk 8]
```diff
@@ -33,29 +45,28 @@ def forward_gpu(self, x):
...
+        for i in ind:
+            cdimy = max(0, min(i, self.cdimx) - prev_i)
             s = list(xshape)
             s[self.axis] = cdimy
             y = cuda.empty(s, dtype=x[0].dtype)
-            if cdimy != 0:
-                kernel(y, x[0], cdimy, self.cdimx, self.rdim, bi)
-            bi = i
+            if cdimy == 0:
+                raise ValueError('Not support if shape contains 0')
```
Comment (delta2323) [inline chainer/functions/split_axis.py]: Same as the above.

**Comment by delta2323** [comment]
Please add the test case where at least one of the output tensors has shape that contains 0. In this case , `forward` should raise `ValueError`.

**Comment by okuta** [comment]
I separated `ValueError`.

**Comment by delta2323** [comment]
LGTM :+1:

**Comment by delta2323** [comment]
Good job! As we discussed above, please open new PR about zero-length output.

**Pull request merged**